src/target/riscv/riscv.c

```diff
@@ -2547,16 +2556,29 @@ static int resume_prep(struct target *target, int current,
 	assert(target->state == TARGET_HALTED);
 	RISCV_INFO(r);
 
+	riscv_reg_t dcsr;
+	if (riscv_reg_get(target, &dcsr, GDB_REGNO_DCSR) != ERROR_OK)
```

Comment:
This read of `dcsr` is unnecessary.  You could avoid it by setting `r->need_single_step` when halt reason is investigated and the hart is halted due to an `ebreak` (look into `set_debug_reason()`).

Reply:
addressed

Reply:
Please, consider always setting `r->need_single_step` in `set_debug_reason()`.